### ISSUE TITLE
Can remove the "@types/react from" "@jupyterlab/apputils" dependencies ?

### ISSUE THREAD

**Issue by jiamicu** [jupyterlab/jupyterlab#5456]
**Describe the bug**
Hi! I want to develop a jupyterlab extension base the npm package "antd" . But after I add it the the extension dependencies,  the tsc command will get error.  I find that "antd" require "@types/react" and when the npm install it , npm will also add it to the "@jupyterlab/apputils" folder ,this will cause TS error. 

**To Reproduce**

1. build a new extension with `cookiecutter https://github.com/jupyterlab/extension-cookiecutter-ts`
2. run npm install antd --save
3. npm run build
4. error TS2717: Subsequent property declarations must have the same type. 

**Desktop (please complete the following information):**

- OS: [e.g. macos]
- Browser [e.g. chrome 69.0]
- JupyterLab [e.g. 0.34.2]

**Comment by ian-r-rose** [comment]
Hi @jiamicu, thanks for this. We unfortunately have to export the react typings from that package since they are part of the *public* API of the VDomRenderer. This can cause problems for downstream extension authors, as you describe. The solution is to force the react typings in your extension to be the same as those in apputils using a resolutions key in your package.json. For instance, this is what is done in `@jupyterlab/toc`:

https://github.com/jupyterlab/jupyterlab-toc/blob/4c90060fbbc728f7545c88966b379afdd01eaf9f/package.json#L72-L74

See also #4367, 4386, and https://github.com/jupyterlab/jupyterlab/pull/5268#discussion_r215307531 .
Can you let us know if that solves your problem?

**Comment by jiamicu** [comment]
Hi @ian-r-rose , thank you for your advice. It works for me. 
The way I try, maybe this will help anyone else :
 - add `resolutions` key in package.json
 - use `jlpm install` to install dependences 
 - run `jlpm run build` ,and everything works fine.

**Comment by vidartf** [comment]
@ian-r-rose Would it not be possible to simply export the typings that are part of the API? E.g.

```ts
import * as React from 'react';

// API includes the following types, so we export them:
export {
  ReactElement, // others?
} from 'react';
```

**Comment by blink1073** [comment]
We'd have to re-define those interfaces locally, recursively...

**Comment by blink1073** [comment]
ohhh, I see what you mean.

**Comment by blink1073** [comment]
That could maybe work...

**Comment by vidartf** [comment]
As far as I can tell, it will work as long as the things that need to be exported are pure typings.

**Comment by blink1073** [comment]
Hmm, I think you'd end up with `@types/react` as a hidden peer dependency, since you could end up with an interface mismatch.

**Comment by vidartf** [comment]
If someone is developing with react, they should have `@types/react` as a dev dependency. The question is how to:
- Ensure that any extensions match the react version (I think this is handled by setting react as a singleton package? Or something like this). This solution should work for both JS and TS.
- Avoid pulling in `@types` as a runtime dependency.

If for some reason the typings the extension is developing with locally does not match the typings that lab was built against, the question is whether they will mismatch in such a way to give a compilation error. If react is following semver correct, I don't think that should be an issue.

**Comment by vidartf** [comment]
(Note, I did some edits to clean up my previous post)

**Comment by blink1073** [comment]
I don't think it is a matter of react following semver, but `@types/react` following semver, which is less likely.

**Comment by blink1073** [comment]
And yes, I've seen this manifest as a compiler error where a deeply nested React interface has a mismatch.

**Comment by vidartf** [comment]
> And yes, I've seen this manifest as a compiler error where a deeply nested React interface has a mismatch.

Ouch. But, the solution for that would be the same as the current solution, right (i.e. to constrain the typings version to match that of lab)? So then I guess this becomes a question of whether those interfaces will mismatch rarely enough for this to be a better solution or not. And I guess which of the error messages are more helpful. I guess the current error message is at least somewhat helpful.

Will there be any difference between the two solution in terms of maintenance overhead? Both from the perspective of lab, and for extension maintainers. If not, then there's no reason to switch.

**Comment by ian-r-rose** [comment]
One of the problems I saw the last time I did a deep dive on this was that if we *don't* declare  `@types/react` as a normal dependency, then compilation of packages that include `@jupyterlab/apputils` but are not using react fails. That is to say, the compiler will look for react typings, not find them, and fail, making `@types/react` effectively a hidden peer dependency (as @blink1073 suggests).

Also, `@types/react` doesn't really follow semver in the same way that a code package does. I found that having deduplicating of that package was not very reliable, and having extra versions around tended to cause conflicts. I have note experimented with @vidartf's idea of only exporting the interfaces we use, but I suppose that I would worry about the same kinds of type conflicts.
